### PR TITLE
Support saving/loading state data using an HTTP URL

### PR DESCRIPTION
**Pull request by divolgin**
What I Did
------------
Added `url` value for `state-from` command line argument.

How I Did it
------------
Implemented the support.

How to verify it
------------
Run ship with new arguments

```
ship init --state-from=url --state-put-url http://somehost/path --state-get-url http://somehost/path
```

Description for the Changelog
------------
Command line argument `state-from` can have value `url`

Picture of a Ship (not required but encouraged)
------------

### CODE REVIEW

File: pkg/state/manager.go [hunk 1]
```diff
@@ -50,6 +43,11 @@ type Manager interface {
 	AddCA(name string, newCA util.CAType) error
 }
 
+type StateSerializer interface {
```
Comment (laverya) [inline pkg/state/manager.go]: Does this need to be public? It does not appear to be used elsewhere

Comment (divolgin) [reply pkg/state/manager.go]: Nope. Does not need to be public at all.

File: pkg/state/state_secret.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	if err != nil {
+		return State{}, errors.Wrap(err, "get in cluster config")
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return State{}, errors.Wrap(err, "get kubernetes client")
+	}
+
+	if s.secretNS == "" {
```
Comment (dexhorthy) [inline pkg/state/state_secret.go]: it looks like the secret serializer checks if all required parameters are set. Do we check that the get/put URLs are provided when state-from=url ?

Comment (divolgin) [reply pkg/state/state_secret.go]: I kind of debated this.  There is no viper object here, so it's not really appropriate to hardcode command line arguments here.  And you still get a pretty meaningful error if the URL is an empty string.

Comment (dexhorthy) [reply pkg/state/state_secret.go]: I think agree: in either case, doing validation inside the serializer implementation seems like the wrong place. Feels like if we add validation, it should probably be a bit higher up, where we pull stuff out of viper

**Pull request merged**